### PR TITLE
Refactor spaces and subnets from networkingcommon

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Work done in preparation of upcoming changes to subnet and space commands.  Start cleanup of
apiserver side with refactor of spaces and subnets from networkingcommon.

Networkingcommon/spaces moved to the spaces client facade. Networkingcommon/subnets moved to the subnets client facade. No need for this functionality in common, not used by more than one facade.

BackingSpace, BackingSubnet and BackupSubnetInfo still used in stub_network for tests, leave in networkingcommon until tests can be rewritten.

Test duplicated between facade and networking common removed.

## QA steps

No change in behavior is expected.

### CODE REVIEW

File: apiserver/common/networkingcommon/shims.go [hunk 1]
```diff
@@ -48,83 +51,3 @@ func (s *spaceShim) Subnets() ([]BackingSubnet, error) {
 	}
 	return subnets, nil
 }
-
-func NewStateShim(st *state.State) (*stateShim, error) {
```
Comment (hmlanigan) [inline apiserver/common/networkingcommon/shims.go]: no longer used.

File: apiserver/facades/agent/proxyupdater/proxyupdater_test.go [hunk 2]
```diff
@@ -37,7 +36,6 @@ var _ = gc.Suite(&ProxyUpdaterSuite{})
 
 func (s *ProxyUpdaterSuite) SetUpSuite(c *gc.C) {
 	s.BaseSuite.SetUpSuite(c)
-	s.StubNetwork.SetUpSuite(c)
```
Comment (hmlanigan) [inline apiserver/facades/agent/proxyupdater/proxyupdater_test.go]: drive by removal of used setup in test.

File: apiserver/facades/client/subnets/subnets_test.go [hunk 3]
```diff
@@ -583,3 +583,339 @@ func (s *SubnetsSuite) TestListSubnetsAllSubnetError(c *gc.C) {
 	_, err := s.facade.ListSubnets(params.SubnetsFilters{})
 	c.Assert(err, gc.ErrorMatches, "no subnets for you")
 }
+
+func (s *SubnetsSuite) TestListSubnetsBlankSpaceTag(c *gc.C) {
```
Comment (hmlanigan) [inline apiserver/facades/client/subnets/subnets_test.go]: The only two tests from networking common not duplicates.

File: apiserver/facades/client/spaces/spaces.go [hunk 4]
```diff
@@ -249,3 +312,16 @@ func (api *API) ReloadSpaces() error {
 	}
 	return errors.Trace(api.backing.ReloadSpaces(env))
 }
+
+// supportsSpaces checks if the environment implements NetworkingEnviron
+// and also if it supports spaces.
+func (api *API) supportsSpaces() error {
```
Comment (achilleasa) [inline apiserver/facades/client/spaces/spaces.go]: Small nitpick: `supportsSpaces` reads like a function that one would expect to return a bool value. Perhaps something like `checkSupportForSpaces` or `ensureSpacesSupport` would read better.

Feel free to ignore this comment.

File: apiserver/facades/client/subnets/shims.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	"gopkg.in/juju/names.v3"
+
+	"github.com/juju/juju/apiserver/common/networkingcommon"
+	"github.com/juju/juju/core/network"
+	providercommon "github.com/juju/juju/provider/common"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/stateenvirons"
+)
+
+func NewStateShim(st *state.State) (*stateShim, error) {
```
Comment (achilleasa) [inline apiserver/facades/client/subnets/shims.go]: Missing comment.

File: apiserver/facades/client/subnets/subnets.go [hunk 6]
```diff
@@ -7,26 +7,58 @@ import (
...
+	"github.com/juju/juju/environs"
 	"github.com/juju/juju/environs/context"
 	"github.com/juju/juju/permission"
+	providercommon "github.com/juju/juju/provider/common"
 	"github.com/juju/juju/state"
 )
 
+var logger = loggo.GetLogger("juju.apiserver.subnets")
+
+type Backing interface {
```
Comment (achilleasa) [inline apiserver/facades/client/subnets/subnets.go]: Missing comment.

File: apiserver/facades/client/spaces/spaces_test.go [hunk 7]
```diff
@@ -212,12 +216,64 @@ func (s *SpacesSuite) TestAddSpacesManySubnets(c *gc.C) {
...
+	s.checkAddSpaces(c, p)
+}
+
 func (s *SpacesSuite) TestAddSpacesAPIError(c *gc.C) {
 	apiservertesting.SharedStub.SetErrors(
 		nil,                                // Backing.ModelConfig()
 		nil,                                // Backing.CloudSpec()
 		nil,                                // Provider.Open()
-		nil,                                // ZonedNetworkingEnviron.SupportsSpaces()
+		nil,                                // ZonedNetworkingEnviron.supportsSpaces()
```
Comment (manadart) [inline apiserver/facades/client/spaces/spaces_test.go]: Looks like GoLand refactoring has changed these comments incorrectly.

File: apiserver/facades/client/subnets/cache.go [hunk 8]
```diff
@@ -22,12 +21,10 @@ import (
 	providercommon "github.com/juju/juju/provider/common"
 )
 
-var logger = loggo.GetLogger("juju.apiserver.common.networkingcommon")
-
 // addSubnetsCache holds cached lists of spaces, zones, and subnets, used for
 // fast lookups while adding subnets.
 type addSubnetsCache struct {
```
Comment (manadart) [inline apiserver/facades/client/subnets/cache.go]: Are we going to ditch this too at some point?

Comment (hmlanigan) [reply apiserver/facades/client/subnets/cache.go]: Yes.  Trying to make the changes in stages.  First is to refactor, second to get rid of the cache, last is to change to mock tests.

File: apiserver/facades/client/spaces/spaces_test.go (hunk 7 above)
Comment (hmlanigan) [reply apiserver/facades/client/spaces/spaces_test.go]: fixed, ta

File: apiserver/common/networkingcommon/shims.go [hunk 9]
```diff
@@ -5,17 +5,16 @@ package networkingcommon
...
 	"github.com/juju/juju/core/life"
-	corenetwork "github.com/juju/juju/core/network"
-	providercommon "github.com/juju/juju/provider/common"
 	"github.com/juju/juju/state"
-	"github.com/juju/juju/state/stateenvirons"
 )
 
 // NOTE: All of the following code is only tested with a feature test.
 
+func NewSubnetShim(sub *state.Subnet) *subnetShim {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply apiserver/common/networkingcommon/shims.go]: added

File: apiserver/common/networkingcommon/shims.go [hunk 10]
```diff
@@ -32,6 +31,10 @@ func (s *subnetShim) Status() string {
 	return "in-use"
 }
 
+func NewSpaceShim(sp *state.Space) *spaceShim {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply apiserver/common/networkingcommon/shims.go]: add

File: apiserver/facades/client/spaces/shims.go [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/juju/errors"
+	"gopkg.in/juju/names.v3"
+
+	"github.com/juju/juju/apiserver/common/networkingcommon"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/stateenvirons"
+)
+
+func NewStateShim(st *state.State) (*stateShim, error) {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply apiserver/facades/client/spaces/shims.go]: add

File: apiserver/facades/client/spaces/spaces.go [hunk 12]
```diff
@@ -27,6 +29,25 @@ type BlockChecker interface {
 	RemoveAllowed() error
 }
 
+type Backing interface {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply apiserver/facades/client/spaces/spaces.go]: add

File: apiserver/facades/client/spaces/spaces.go (hunk 4 above)
Comment (hmlanigan) [reply apiserver/facades/client/spaces/spaces.go]: changed to checkSupportsSpaces().

File: apiserver/facades/client/subnets/shims.go (hunk 5 above)
Comment (hmlanigan) [reply apiserver/facades/client/subnets/shims.go]: added

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**